### PR TITLE
plat/xilinx: Add support for new platform management APIs for ZynqMP

### PR DESCRIPTION
**Pull request by jollysxilinx**
This patch adds support for new APIs to communication layer with firmware.
APIs include clock control, pinmux control, query and dev ioctl.

QUERY api provide interface to query soc specific data from ATF
IOCTLs provide interface to support different device configurations
Clock control APIs enables support for clock control using PMU(platform management unit)
Pinmux control APIs enables support for pin control using PMU.(platform management unit)

Changes from #1230:
    -Added pinmux query api support
    -Fixed memory map for DEBUG build

Signed-off-by: Jolly Shah jollys@xilinx.com

### CODE REVIEW

**Comment by davidcunado-arm** [comment]
jenkins: test this please

**Comment by jollysxilinx** [comment]
Hi @davidcunado-arm ,

May I know failures?

**Comment by davidcunado-arm** [comment]
@jollysxilinx 
The failures are in the static analysis - there are quite a few (>400), but some of these are likely to be false positives or the same code causing multiple issues; I'll try to distill them to a more reasonable set.

**Comment by davidcunado-arm** [comment]
jenkins: test this please

**Comment by davidcunado-arm** [comment]
jenkins: test this please

**Comment by davidcunado-arm** [comment]
jenkins: test this please

File: plat/xilinx/zynqmp/include/platform_def.h [hunk 1]
```diff
@@ -34,9 +34,14 @@
  * little space for growth.
  */
 #ifndef ZYNQMP_ATF_MEM_BASE
+#if !defined(DEBUG)
```
Comment (davidcunado-arm) [inline plat/xilinx/zynqmp/include/platform_def.h]: Can you check that this is working as you expect? I believe DEBUG is always defined - perhaps should be #if DEBUG?

Comment (sivadur) [reply plat/xilinx/zynqmp/include/platform_def.h]: Hi,
Yes, it doesn't work, we will be correct it with if !DEBUG instead.

**Comment by jollysxilinx** [comment]
Changes done and pushed on same branch. Please review.

**Comment by davidcunado-arm** [comment]
@jollysxilinx 
Apologies - we're about to complete some release activities; I'll look at your once we've completed them. Sorry for the delay!

**Comment by davidcunado-arm** [comment]
jenkins: test this please

**Comment by davidcunado-arm** [comment]
@jollysxilinx 
We're getting build failures:

aarch64-linux-gnu-ld.bfd: build/zynqmp/release/bl31/bl31.elf section `coherent_ram' will not fit in region `RAM'
aarch64-linux-gnu-ld.bfd: BL31 image has exceeded its limit.
aarch64-linux-gnu-ld.bfd: region `RAM' overflowed by 1 bytes
make: *** [build/zynqmp/release/bl31/bl31.elf] Error 1
Build failed with code 2

For the following builds:
- Build options: CROSS_COMPILE=aarch64-linux-gnu- V=1 PLAT=zynqmp RESET_TO_BL31=1 SPD=tspd
- Build options: CROSS_COMPILE=aarch64-linux-gnu- V=1 PLAT=zynqmp RESET_TO_BL31=1 SPD=tspd ZYNQMP_TSP_RAM_LOCATION=tdram

Otherwise, looks ok - I think we're on diminishing returns on trying to resolve all the MISRA issues; remaining ones are predominantly Advisory.

**Comment by sivadur** [comment]
@jollysxilinx I think we need to get rid of SPD flag for now and update zynqmp read me accordingly.

**Comment by dp-arm** [comment]
jenkins: test this please

**Comment by jollysxilinx** [comment]
Hi David,

As discussed, We are working on different memory options to accommodate higher footprint. Once we finalize, I can submit a separate patch to fix memory issues. Please accept this patchset if rest looks ok.

Thanks,
Jolly Shah

**Comment by dp-arm** [comment]
jenkins: test this please

**Comment by dp-arm** [comment]
> As discussed, We are working on different memory options to accommodate higher footprint. Once we finalize, I can submit a separate patch to fix memory issues. Please accept this patchset if rest looks ok.

I am doing a CI build and will merge soon.

**Comment by dp-arm** [comment]
jenkins: test this please

**Pull request merged**